aiida_siesta/workflows/utils/converge_absclass.py

```diff
@@ -0,0 +1,307 @@
+        '''
+        Takes care of returning the results of the convergence to the user
+        '''
+
+        converged = Bool(self.converged)
+        converged_index = Int(getattr(self.ctx, 'converged_index', -1))
+        iteration_keys = List(list=list(self.ctx.iteration_keys))
+        variable_values = List(list=self.ctx.variable_values)
+        target_values = List(list=self.ctx.target_values)
+
+        outputs = generate_convergence_results(
+            iteration_keys, variable_values, target_values, converged, converged_index
+        )
+
+        if converged:
+            self.report(
+                '\n\nConvergence has been reached! Converged parameters:'
+                f'{outputs["converged_parameters"].get_dict()}\n'
+            )
+        else:
+            self.report('\n\nWARNING: Workchain ended without finding convergence\n ')
+
+        self.out_many(outputs)
+
+        super().return_results()
+
+
+class SequentialConverger(ProcessInputsIterator):
+    '''
+    Launches several convergence workchains sequentially. It is actally an iterator that
+    iterate a _process_class which is a converger!!!!
+
+    It is perfectly fine to define a class:
+        class SiestaSeqConv(ProcessInputsIterator):
+            _process_class = SiestaConverger
+            _expose_inputs_kwargs = {"namespace": 'converger_inputs'}
+    The presence of a namespace in _expose_inputs_kwargs is necessary to avoid conflict between the exposed
+    iterate_over of _process_class  and the iterate_over of the SequentialConverger itself.
+    However the class above will have a very convoluted way to submit calculations
+        iterate_over = {"iterate_over" : [{"meshcutoff" : [...]},{"paoenergyshift":[....]}]
+        inputs = { "converger_inputs" : {"inp1": ..., "inp2": ..., ... }  }
+        submit(SiestaSeqConv, **inputs, iterate_over=iterate_over)
+    and it will miss the part of incorporating the already converged "meshcutoff" into the convergence
+    of "paoenergyshift".
+    Here comes the need for a more complex class that facilitate the task of running sequential convergence.
+
+    The iterate_over input is modified to accept a List where each item is a dict of accepted iterate_over
+    keyword of _process_class (The Converger!!!). Imposes the modifications of the port and
+    _iterate_input_serializer.
+
+    THIS CLASS CAN NOT BE USED DIRECTLY, you need to subclass it and specify a cls._process_class,
+    which should be the converger that you want to use.
+    '''
+
+    _expose_inputs_kwargs = {'exclude': ("iterate_over",), "namespace": 'converger_inputs'}
+    _reuse_inputs = True
+
+    def __init__(self, *args, **kwargs):
+        """Construct the instance. Just needed to implement a check on the passed _process_class"""
+
+        try:
+            if not issubclass(self._process_class, BasicConverger):
+                raise ValueError('no valid `_process_class` attribute, it must be a Converger!')
```

Comment:
This is a good idea. I will look better into it.

Reply:
I'm having a look and this `__init_subclass__` was introduced in python 3.6, I guess this is the reason why it is not used in aiida-core which still supports python 3.5. However we could use it. I will use it also in ProcessInputsIterator to impose that `_process_class` is not None or a random thing. It must be a process. Ok?

Reply:
> I'm having a look and this __init_subclass__ was introduced in python 3.6.

I didn't know that, good to know! Then we could use a metaclass to get the same behavior.

> I will use it also in ProcessInputsIterator to impose that _process_class is not None or a random thing.

Although now that you say this, I realise that maybe they do it on `__init__` so that you can still create a "helper class" that is not meant to be used. I mean something like:

```python
class CustomIterator(Iterator):
    ...implement extra functionality
    # This class would not be possible to create if we check on class definition

class MyIterator(CustomIterator):
    _process_class = Whatever
   # This is the class that is really meant to be used.
```

So I would say we either check in class definition and allow `None`, or we check on `__init__`. Probably it's better on `__init__` as you did to ensure that the real workchains will work. But still, maybe it's better to write the code only once in the base iterator and make it tunable with a `_valid_processes` variable or something like that. I mean:

```python
class GeneralIterator(Workchain):

    _process_class = None
    _valid_processes = Process

    def __init__(self, *args, **kwargs):
        """Construct the instance. Just needed to implement a check on the passed _process_class"""
        from aiida.engine import Process  # pylint: disable=cyclic-import

        try:
            if not issubclass(self._process_class, self._valid_processes):
                raise ValueError(f"{self.__class__.__name__}'s '_process_class' variable must be a subclass of {self._valid_processes}, but {self._process_class} was provided instead")
        except TypeError:
            raise ValueError(f"{self.__class__.__name__}'s '_process_class' variable must be a subclass of {self._valid_processes}, but {self._process_class} was provided instead")

        super().__init__(*args, **kwargs)
```
Then we can do 

```python
class SequentialConverger(GeneralIterator):
    _valid_processes = BasicConverger
```

And we have this problem solved for any new iterator that requires a specific process without needing to repeat the code. What do you think?

Reply:
I see the problem. For the moment I wouldn't complicate too much the situation. I would just allow `_process_class` to be None or a Process in `__init_subclass__`. Then I would impose that   `_process_class` is not None in `__init__`. Let's see the comments we receive.

Reply:
But then you need to overwrite `__init_subclass__` in `SequentialConverger`, right?

Reply:
Yes, did in last commit